src/gg.regression.unity.bots/Editor/Scripts/ActionManager/RGActionManagerUI.cs

```diff
@@ -237,22 +236,613 @@ private void UpdateGUI()
             _analyzeBtn.SetEnabled(!EditorApplication.isPlayingOrWillChangePlaymode);
         }
 
+        private void SetPropertyOverride(RGActionPropertyInstance prop, object newValue)
+        {
+            prop.SetValue(newValue);
+            RGActionManager.Settings.StoreProperty(prop);
+            RGActionManager.SaveSettings(RGActionManager.Settings);
+        }
+
+        private void ResetProperty(RGActionPropertyInstance prop, object origValue)
+        {
+            prop.SetValue(origValue);
+            RGActionManager.Settings.ResetProperty(prop);
+            RGActionManager.SaveSettings(RGActionManager.Settings);
+        }
+
+        private IEnumerable<VisualElement> EnumerateDescendents(params VisualElement[] elems)
+        {
+            foreach (var elem in elems)
+            {
+                yield return elem;
+                foreach (var child in elem.Children())
+                {
+                    foreach (var desc in EnumerateDescendents(child))
+                    {
+                        yield return desc;
+                    }
+                }
+            }
+        }
+
+        private VisualElement CreateControlForProperty(RGActionPropertyInstance prop, object origValue, Action valueChangedCallback, out Action onPropReset)
+        {
+            const int numericFieldWidth = 60;
+            var valueType = prop.GetValue().GetType();
+            if (valueType == typeof(string))
+            {
+                TextField textField = new TextField();
```

Comment:
All of this UI code reminds me of the days before React and Angular on the web lol

Reply:
Unity does have the UXML layout language, but pretty much everything in the action manager UI is dynamically constructed from the analysis outcome, so just constructing the components in code was easier.